### ISSUE TITLE
Possible issue with NL swallowing

### ISSUE THREAD

**Issue by YourPsychiatrist** [yshavit/antlr-denter#22]
Hello! Thank you for creating this little library. It looks too convenient. Sadly I am having troubles getting it to work. Here is a TL;DR of my grammar file:

```antlr
grammar Scribble;

// copied from https://github.com/yshavit/antlr-denter/blob/main/Java/examples/src/main/resources/com/yuvalshavit/antlr4/examples/benchmark/BenchGramDenting.g4
// but adapted so that the class names match up
tokens { INDENT, DEDENT }
@lexer::header {
  import com.yuvalshavit.antlr4.DenterHelper;
}
@lexer::members {
  private final DenterHelper denter = new DenterHelper(NL, ScribbleParser.INDENT, ScribbleParser.DEDENT) {
    @Override
    public Token pullToken() {
      return ScribbleLexer.super.nextToken();
    }
  };

  @Override
  public Token nextToken() {
    return denter.nextToken();
  }

  public Token rawNextToken() {
    return super.nextToken();
  }
}

conditional : 'if' IDENTIFIER 'then' INDENT DIGIT+ DEDENT ;

NL: ('\r'? '\n' ' '*) ;
DIGIT : [0-9] ;
IDENTIFIER : [a-zA-Z] [a-zA-Z0-9_]* ;
// ignore spaces and tabs
WS : [ \t] -> skip ;
```

Now when I try feeding it the following sample input:

```text
if a then
  1

```

(which is concretely: `if a then\n  1\n<EOF>`) it fails. Am I missing something?

In IntelliJ using the ANTLR preview tool, I get this result:

![image](https://user-images.githubusercontent.com/13717708/125298010-1080b700-e328-11eb-8ee8-571c381da2f4.png)

Initially, I tried to directly incorporate it into an existing, larger grammar, but after I couldn't get it to work I essentially reduced it to the example above to make sure I am trying the simplest possible thing. Also: I tried the `BenchGramDenting` sample grammar + source and it appeared to work fine. Do you have any first guesses? Anything that I obviously overlooked?

**Comment by YourPsychiatrist** [comment]
As a short note, changing the rule to `conditional : 'if' IDENTIFIER 'then' INDENT DIGIT+ NL DEDENT ;` seems to fix at least the errors printed to stderr, but the IntelliJ plugin still complains.. so I am thinking its a bit misleading. Still curious why i need the `NL` before the `DEDENT`, eventhough you didn't use it in your examples :thinking:

**Comment by yshavit** [comment]
Hi there! I can take a look tonight (US/Eastern time), but in the meanwhile: what happens if you move the `WS : [ \t] -> skip` up, to be right after the `NL` rule? It's been a while since I worked with this library to be honest, but if I recall, antlr4 does use the rule orders for some disambiguation at times.

```
NL: ('\r'? '\n' ' '*) ;
// ignore spaces and tabs
WS : [ \t] -> skip ;

DIGIT : [0-9] ;
IDENTIFIER : [a-zA-Z] [a-zA-Z0-9_]* ; 
```

**Comment by YourPsychiatrist** [comment]
Yeah AFAIK antlr does test for the earlier rules first but no matter how I move the rules around I don't get the desired effect :/ Thanks for the quick reply tho

**Comment by YourPsychiatrist** [comment]
I also wanna mention that I was sloppy earlier, I can't get your benchdent example to run either :/ Both console and GUI give me errors.

**Comment by yshavit** [comment]
Alright, thanks for that info! I'll try to take a look tonight.

Just so we fix some variables: can you share what version of Java and IDE you're using?

**Comment by YourPsychiatrist** [comment]
Much obliged

The IDE is IntelliJ IDEA 2021.2 Beta (#IU-212.4638.7)
`java -version` gives me
```text
openjdk version "16.0.1" 2021-04-20
OpenJDK Runtime Environment GraalVM CE 21.1.0 (build 16.0.1+9-jvmci-21.1-b05)
OpenJDK 64-Bit Server VM GraalVM CE 21.1.0 (build 16.0.1+9-jvmci-21.1-b05, mixed mode, sharing)
```

For what it's worth, I am also using Scala 3 with sbt 1.5.4. I think that shouldn't have too much of an impact tho.

**Comment by YourPsychiatrist** [comment]
Do you happen to have had the chance to look into it yet? No pressure ofc

**Comment by yshavit** [comment]
Hi there, sorry for the delay -- last weekend and the week ended up getting a bit crazy.

I looked into it, and I think there was a bug in the benchmark grammar -- sorry! You do need the `NL` token after your digits:

```
conditional : 'if' IDENTIFIER 'then' INDENT DIGIT+ NL DEDENT ;
```

This will parse something like:

```
if a then
  1 2 3
```

... but _not_:

```
if a then
  1
  2
  3
```

If you want the latter, you need to instead have the newline after each digit:

```
conditional : 'if' IDENTIFIER 'then' INDENT (DIGIT NL)+ DEDENT ;
```

This behavior is (correctly) described [in the readme](https://github.com/yshavit/antlr-denter#using-the-tokens-in-your-parser):

>  One bit worth noting is that when the denter injects DEDENT tokens, it'll prefix any string of them with a single `NL`. ... This is done so that simple expressions can be terminated by the NL token without worrying about surrounding context (an impending dedent or EOF). In this case, `universe` and `dolly` represent simple expressions, and you can imagine that the grammar would contain something like statement: expr NL | helloBlock;

In other words, the problem with your grammar is that you essentially defined the digit as an expression instead of a statement: you need to add the newline so that it's a statement. While this may feel odd, it means that expressions never need the NL (which is good, because it makes compound statements easier!), single-line statements always end in a newline, and block statements (`if`, etc) only need to worry about the INDENT and DEDENT, because the block's statements will gobble up the newlines correctly.

I apologize for the broken example grammar, which lead you astray! If this explanation makes sense, please feel free to resolve the ticket -- otherwise, we can definitely keep chatting. :-) I also just put up PR #23 to fix that example grammar.

**Comment by YourPsychiatrist** [comment]
Thank you very much for the elaborate response! I will give that a try and report back :)

**Comment by SkittlePox** [comment]
I've also been following this issue and adding the `NL` tokens to the end of single-line expressions seems to do the trick when testing with `pygrun`. However, as @YourPsychiatrist mentioned above, the JetBrains ANTLR plugin does not play nicely and gives errors. Mine gives a `mismatched input '\n' expecting ...` error. My guess is that it's probably an issue with the plugin, not antlr-denter

**Comment by YourPsychiatrist** [comment]
@SkittlePox Absolutely. I resorted to printing to the console for now and investigating the string tree. I will write a little thing that spits out DOT code for graphviz probably. I am thinking that the ANTLR plugin doesn't recognize the library, but that's a wild guess.

**Comment by YourPsychiatrist** [comment]
Closing this issue since it has been resolved.

**Comment by yshavit** [comment]
Glad to hear the new grammar works!

Sorry the intellij plugin doesn't -- that stinks. :-\ Unfortunately, I don't actually use this project anymore (I gave up on the pet project that I was using it for), so I haven't used the plugin in quite some time, and I can't provide much help with it.

**Comment by yshavit** [comment]
Oh you know what -- this sounded familiar, so I searched a bit. @SkittlePox is right that this is an issue with the plugin, and it looks like it's due to technical limitations in the plugin: antlr/intellij-plugin-v4#137.

I did say in this (my) repo's #5 that the plugin works well, but I can't remember what my definition of "well" was. ;-) And I definitely don't remember specifics.
